src/Domain/Structure/ElementSearchTree.hpp

```diff
@@ -0,0 +1,100 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/geometry.hpp>
+#include <boost/geometry/index/rtree.hpp>
+#include <cstddef>
+#include <map>
+#include <utility>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Structure/ElementId.hpp"
+
+namespace bg = boost::geometry;
+
+namespace boost::geometry::traits {
+// Make Tensor compatible with Boost.Geometry
+template <typename DataType, size_t Dim, typename Frame>
+struct tag<tnsr::I<DataType, Dim, Frame>> {
+  using type = bg::point_tag;
+};
+template <typename DataType, size_t Dim, typename Frame>
+struct coordinate_type<tnsr::I<DataType, Dim, Frame>> {
+  using type = DataType;
+};
+template <typename DataType, size_t Dim, typename Frame>
+struct coordinate_system<tnsr::I<DataType, Dim, Frame>> {
+  using type = bg::cs::cartesian;
+};
+template <typename DataType, size_t Dim, typename Frame>
+struct dimension<tnsr::I<DataType, Dim, Frame>>
+    : std::integral_constant<size_t, Dim> {};
+template <typename DataType, size_t Dim, typename Frame, size_t GetDimension>
+struct access<tnsr::I<DataType, Dim, Frame>, GetDimension> {
+  static constexpr DataType get(const tnsr::I<DataType, Dim, Frame>& point) {
+    return ::get<GetDimension>(point);
+  }
+  static void set(tnsr::I<DataType, Dim, Frame>& point, const DataType& value) {
+    ::get<GetDimension>(point) = value;
+  }
+};
+// Make ElementId compatible with Boost.Geometry
+template <size_t Dim>
+struct tag<ElementId<Dim>> {
+  using type = bg::box_tag;
+};
+template <size_t Dim>
+struct coordinate_type<ElementId<Dim>> {
+  using type = double;
+};
+template <size_t Dim>
+struct coordinate_system<ElementId<Dim>> {
+  using type = bg::cs::cartesian;
+};
+template <size_t Dim>
+struct dimension<ElementId<Dim>> : std::integral_constant<size_t, Dim> {};
+template <size_t Dim>
+struct point_type<ElementId<Dim>> {
+  using type = tnsr::I<double, Dim, ::Frame::BlockLogical>;
+};
+template <size_t Dim, size_t Index, size_t GetDimension>
+struct indexed_access<ElementId<Dim>, Index, GetDimension> {
+  static constexpr double get(const ElementId<Dim>& element_id) {
+    if constexpr (Index == 0) {
+      return element_id.segment_id(GetDimension).endpoint(Side::Lower);
+    } else {
+      return element_id.segment_id(GetDimension).endpoint(Side::Upper);
+    }
+  }
+};
+}  // namespace boost::geometry::traits
+
+namespace domain {
+
+template <size_t Dim>
+using ElementSearchTree =
+    bg::index::rtree<ElementId<Dim>, bg::index::quadratic<16>>;
+
+/*!
+ * \brief Sorts element IDs into a search tree for efficient searching.
```

Comment:
I have no idea what the input or output is. What is the size_t in the map? Why a map? It's not at all clear to me from this documentation how this would change lookup time of some spatial coordinate (x,y,z).